### PR TITLE
ONNXHybridTransformPass

### PR DESCRIPTION
**Pull request by sorenlassen**
Includes a new pattern-based shape inference implementation.

So far only combines shape inference and canonicalization but the plan is to include constant propagation and decomposition so that a single pass with all the patterns can cascade shape inference and all the transforms in a single pass.

I encountered a model with a long dependency chain between shape inference, canonicalization, and constant propagation which requires an incredible number of repetitive passes with the current pass infrastructure.

See PR #1770 for an earlier version of this PR.

### CODE REVIEW

**Comment by sorenlassen** [comment]
If I enable the new pass, by changing the `--onnx-hybrid-pass` flag default to true, then these lit tests fail:
```
Failed Tests (8):
  Open Neural Network Frontend :: onnx/onnx_bundle_memory_pool.mlir
  Open Neural Network Frontend :: onnx/onnx_lowering.mlir
  Open Neural Network Frontend :: onnx/onnx_lowering_O3.mlir
  Open Neural Network Frontend :: onnx/onnx_lowering_range_op.mlir
  Open Neural Network Frontend :: onnx/onnx_lowering_with_canonicalize.mlir
  Open Neural Network Frontend :: onnx/onnx_lowering_with_canonicalize_O3.mlir
  Open Neural Network Frontend :: onnx/onnx_seq_dealloc.mlir
  Open Neural Network Frontend :: onnx/onnx_shape_inference.mlir
```
I haven't looked into them yet.

**Comment by chentong319** [comment]
We can progressively increase the patterns in the hybrid pass. It might be too much to add all the patterns at once.
Need help on If/Loop/Scan?
What's the failure for lit test, compilation error or match error?

File: src/InitOMPasses.hpp [hunk 1]
```diff
@@ -38,7 +38,17 @@ void initOMPasses(int optLevel) {
...
   mlir::registerPass([]() -> std::unique_ptr<mlir::Pass> {
-    return createShapeInferencePass();
+    return createONNXHybridTransformPass();
+  });
+
+  mlir::registerPass([]() -> std::unique_ptr<mlir::Pass> {
+    if (enableONNXHybridPass) {
+      // Replaces shape inference in onnx-mlir-opt which is used in all the
+      // lit tests, except the parse lit tests.
+      return createONNXShapeInferenceTransformPass();
```
Comment (chentong319) [inline src/InitOMPasses.hpp:48]: I think that we can add both createONNXHybridTransformPass and createONNXShapeInferenceTransformPass and keep the ShapeInferencePass as it was. We can copy/add new lit test for the new passes without  touch the existing lit test. The ShapeInferencePass will be retired after ONNXShapeInferenceTransformPass becomes stable.

File: src/Transform/ONNX/ONNXHybridTransformPass.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+// the parent if/loop/scan op each accomplishes one half of shape propagation to
+// and from the subgraph: The first run propagates input shapes from the parent
+// op to the subgraph and the second run propagates result shapes from the
+// subgraph to the parent op.
+struct ONNXHybridTransformPass
+    : public PassWrapper<ONNXHybridTransformPass, OperationPass<func::FuncOp>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(ONNXHybridTransformPass)
+
+  ONNXHybridTransformPass(bool shapeInferenceOnly)
+      : shapeInferenceOnly(shapeInferenceOnly) {}
```
Comment (chentong319) [inline src/Transform/ONNX/ONNXHybridTransformPass.cpp:87]: I am think of a  general control on which transformation pattern, not just shape inference only, will be added in the hybrid pass.

**Comment by sorenlassen** [comment]
> Need help on If/Loop/Scan?
> What's the failure for lit test, compilation error or match error?

it's mostly match errors

there's a crash in `@test_loop_simple_one_scan_main_graph` in onnx_shape_inference.mlir,
it crashes because we attempt to cast an unranked tensor type `tensor<*xi64>` to RankedTensorType in Loop.cpp:
https://github.com/onnx/onnx-mlir/blob/main/src/Dialect/ONNX/ONNXOps/ControlFlow/Loop.cpp#L84

maybe it's because I left out the fixes to ONNXScanOp::inferShapes() and ONNXLoopOp::inferShapes() from PR #1770 when I rebased

your help would be appreciated if you can checkout my branch and run the lit tests with the default value of the enableONNXHybridPass flag set to true in CompilerOptions.cpp

**Comment by sorenlassen** [comment]
I should probably also try to port the fixes to lit tests from PR #1770, it looks like some of the same lit tests are the same as I fixed back then

File: src/Compiler/CompilerOptions.cpp [hunk 3]
```diff
@@ -208,6 +208,11 @@ llvm::cl::opt<bool> enableSimdDataLayout("simd-data-layout",
                    "Set to 'true' if you want to enable SIMD optimizations."),
     llvm::cl::init(false), llvm::cl::cat(OnnxMlirOptions));
 
+llvm::cl::opt<bool> enableONNXHybridPass("onnx-hybrid-pass",
+    llvm::cl::desc("Enable ONNX hybrid pass (default=true)\n"
+                   "Set to 'false' if you want to disable ONNX hybrid pass."),
+    llvm::cl::init(true), llvm::cl::cat(OnnxMlirCommonOptions));
```
Comment (sorenlassen) [inline src/Compiler/CompilerOptions.cpp:214]: I temporarily changed the default to `true` to surface the lit test failures that results from changing shape inference to the new algorithm

I also fixed half the lit test failures

File: test/mlir/onnx/onnx_lowering_with_canonicalize.mlir [hunk 4]
```diff
@@ -194,11 +194,7 @@ func.func @test_if_sign(%arg0: tensor<f32>) -> tensor<i32> {
 // CHECK-DAG:         [[VAR_8_:%.+]] = arith.cmpf ogt, [[LOAD_PARAM_0_MEM_1_]], [[LOAD_VAR_0_MEM_1_]] : f32
 // CHECK-DAG:         krnl.store [[VAR_8_]], [[RES_1_]][] : memref<i1>
 // CHECK-DAG:         [[LOAD_RES_1_MEM_:%.+]] = krnl.load [[RES_1_]][] : memref<i1>
-// CHECK-DAG:         [[VAR_10_:%.+]] = scf.if [[LOAD_RES_1_MEM_]] -> (memref<i32>) {
-// CHECK-DAG:           scf.yield [[CONSTANT_2_]] : memref<i32>
-// CHECK-DAG:         } else {
-// CHECK-DAG:           scf.yield [[CONSTANT_1_]] : memref<i32>
-// CHECK-DAG:         }
+// CHECK:             [[VAR_10_:%.+]] = arith.select [[LOAD_RES_1_MEM_]], [[CONSTANT_2_]], [[CONSTANT_1_]] : memref<i32>
```
Comment (sorenlassen) [inline test/mlir/onnx/onnx_lowering_with_canonicalize.mlir:197]: @chentong319 do you have any idea why we the changes in this PR causes this difference in lowering?

File: src/Compiler/CompilerOptions.cpp (hunk 3 above)
Comment (sorenlassen) [reply src/Compiler/CompilerOptions.cpp:214]: I figured out the linux CI build failures, it was another crash on an unranked input, this time when compiling the mnist-8 model from the model zoo. The crash was in FuseAddConvNullBiasPattern in src/Dialect/ONNX/Rewrite.td. I added a check `(HasShapeAndRank:$res)` to bypass the pattern when the type of ONNXAddOp is unranked.

Comment (sorenlassen) [reply src/Compiler/CompilerOptions.cpp:214]: On second thought the better fix is to apply to shape inference patterns before the canonicalization patterns in the hybrid pass. In this way the shapes will be inferred before FuseAddConvNullBiasPattern is tested and the pattern will apply. I made this change and moved the added `(HasShapeAndRank:$res)` check to a standalone PR #2159.

File: test/mlir/onnx/onnx_lowering_with_canonicalize.mlir (hunk 4 above)
Comment (sorenlassen) [reply test/mlir/onnx/onnx_lowering_with_canonicalize.mlir:197]: I came up with a workaround in PR #2168

**Comment by chentong319** [comment]
Comments on the control flow Ops. I will use LoopOp as an example.
There are three part of the inferShape function:
1. Set the type of body arguments with the initial value
2. Use parameter  function pointer on the body region. Originally, the function will traverse the body and call inferShape for each Op.
3. Set type for the returned values

In hybrid transformation:
1. when loop is encountered, inferShape is called with a function pointer doing nothing. So we actually perform step#1 and step#3. The step#3 is not needed yet.
2. When return of the loop is encountered, inferShape for its LoopOp is called with a function pointer doing nothing. In this step, what we really want is just step#3.

After you fixed the unranked tensor issue, such flow with unnecessary work is OK.
I am thinking whether we can get rid of the unnecessary work by separating out the code into entering and leaving the node. In hybrid transformation, we pass a Null function pointer. So the code for loop looks like:
```
ONNXLoopOp::JoinEdges() {
      // set type when control flow join
}

ONNXLoopOp::InferShape(doShapeInference) {
     // set the entry type
    if (doShapeInference != nullptr) {
        doShapeInference(insideRegion); // traverse and infer shape for the region
        joinEdges();
     }
}

For the ReturnOp:
if (it is the return for a LoopOp, loop) {
    loop.JoinEdge();
}

In hybrid transformation, always call : inferShape(nullptr)
```
We can merge this PR without such changes. I just want to check with you whether my understanding of the flow is correct.

**Comment by sorenlassen** [comment]
> We can merge this PR without such changes. I just want to check with you whether my understanding of the flow is correct.

Yes, this is correct.

I'd be very interested in making the type of improvements you suggest as a follow-up.

**Review by chentong319 — APPROVED** [review]
LGTM!

File: src/Compiler/CompilerOptions.cpp (hunk 3 above)
Comment (sorenlassen) [reply src/Compiler/CompilerOptions.cpp]: > I temporarily changed the default to true to surface the lit test failures that results from changing shape inference to the new algorithm

I changed the default back to false.

**Comment by sorenlassen** [comment]
@jenkins-droid test this please

**Comment by sorenlassen** [comment]
@gongsu832 can you tell why MLIR-Windows-CI is failing? https://github.com/onnx/onnx-mlir/pull/2098/checks?check_run_id=12765970850

**Comment by gongsu832** [comment]
You can click on the "Details" -> "1 errors / 0 warnings" -> "Build_onnx_mlir_Windows" and you can see the error.

```
2023-04-15T00:52:03.8473519Z ##[section]Starting: Test onnx-mlir
2023-04-15T00:52:04.1222663Z ==============================================================================
2023-04-15T00:52:04.1223031Z Task         : Command line
2023-04-15T00:52:04.1223182Z Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
2023-04-15T00:52:04.1223473Z Version      : 2.212.0
2023-04-15T00:52:04.1223618Z Author       : Microsoft Corporation
2023-04-15T00:52:04.1223797Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
2023-04-15T00:52:04.1224106Z ==============================================================================
2023-04-15T00:52:07.2654004Z Generating script.
2023-04-15T00:52:07.3125315Z ========================== Starting Command Output ===========================
2023-04-15T00:52:07.3440286Z ##[command]"C:\Windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "D:\a\_temp\3c94cc30-c84c-4a26-987d-d36cccbc85ba.cmd""
2023-04-15T00:52:08.0166916Z **********************************************************************
2023-04-15T00:52:08.0168365Z ** Visual Studio 2019 Developer Command Prompt v16.11.25
2023-04-15T00:52:08.0169042Z ** Copyright (c) 2021 Microsoft Corporation
2023-04-15T00:52:08.0170136Z **********************************************************************
2023-04-15T00:52:09.4024705Z [vcvarsall.bat] Environment initialized for: 'x64'
2023-04-15T00:52:11.9495939Z lit: C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\lit\llvm\config.py:46: note: using lit tools: C:\Program Files\Git\usr\bin
2023-04-15T00:52:13.3561772Z -- Testing: 205 tests, 4 workers --
2023-04-15T00:52:17.4210420Z Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.
2023-04-15T00:52:17.4238521Z FAIL: Open Neural Network Frontend :: krnl/krnl_getref_lowering.mlir (138 of 205)
2023-04-15T00:52:17.4240399Z ******************** TEST 'Open Neural Network Frontend :: krnl/krnl_getref_lowering.mlir' FAILED ********************
2023-04-15T00:52:17.4248321Z Script:
2023-04-15T00:52:17.4251530Z --
2023-04-15T00:52:17.4253805Z : 'RUN: at line 1';   d:\a\1\onnx-mlir\build\release\bin\onnx-mlir-opt.exe -O3 --convert-krnl-to-affine --convert-krnl-to-llvm D:\a\1\onnx-mlir\test\mlir\krnl\krnl_getref_lowering.mlir -split-input-file | d:\a\1\llvm-project\build\bin\filecheck.exe D:\a\1\onnx-mlir\test\mlir\krnl\krnl_getref_lowering.mlir
2023-04-15T00:52:17.4255401Z --
2023-04-15T00:52:17.4256204Z Exit Code: 2
2023-04-15T00:52:17.4256642Z 
2023-04-15T00:52:17.4257285Z Command Output (stdout):
2023-04-15T00:52:17.4257602Z --
2023-04-15T00:52:17.4257896Z $ ":" "RUN: at line 1"
2023-04-15T00:52:17.4259020Z $ "d:\a\1\onnx-mlir\build\release\bin\onnx-mlir-opt.exe" "-O3" "--convert-krnl-to-affine" "--convert-krnl-to-llvm" "D:\a\1\onnx-mlir\test\mlir\krnl\krnl_getref_lowering.mlir" "-split-input-file"
2023-04-15T00:52:17.4260221Z note: command had no output on stdout or stderr
2023-04-15T00:52:17.4261140Z error: command failed with exit status: 3221226356
2023-04-15T00:52:17.4262373Z $ "d:\a\1\llvm-project\build\bin\filecheck.exe" "D:\a\1\onnx-mlir\test\mlir\krnl\krnl_getref_lowering.mlir"
2023-04-15T00:52:17.4263252Z # command stderr:
2023-04-15T00:52:17.4264126Z FileCheck error: '<stdin>' is empty.
2023-04-15T00:52:17.4265222Z FileCheck command line:  d:\a\1\llvm-project\build\bin\filecheck.exe D:\a\1\onnx-mlir\test\mlir\krnl\krnl_getref_lowering.mlir
2023-04-15T00:52:17.4265806Z 
2023-04-15T00:52:17.4266463Z error: command failed with exit status: 2
2023-04-15T00:52:17.4266666Z 
2023-04-15T00:52:17.4266919Z --
2023-04-15T00:52:17.4267013Z 
2023-04-15T00:52:21.7567643Z ********************
2023-04-15T00:52:21.7568179Z Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 
2023-04-15T00:52:21.7593153Z ********************
2023-04-15T00:52:21.7593957Z Failed Tests (1):
2023-04-15T00:52:21.7594649Z   Open Neural Network Frontend :: krnl/krnl_getref_lowering.mlir
2023-04-15T00:52:21.7595304Z 
2023-04-15T00:52:21.7595483Z 
2023-04-15T00:52:21.7595925Z Testing Time: 9.70s
2023-04-15T00:52:21.7598372Z   Unsupported:  72
2023-04-15T00:52:21.7598869Z   Passed     : 132
2023-04-15T00:52:21.7599320Z   Failed     :   1
2023-04-15T00:52:21.7813231Z [1/6] cmd.exe /C "cd /D D:\a\1\onnx-mlir\build\src\Runtime\jni && "C:\Program Files\CMake\bin\cmake.exe" -E copy_if_different D:/a/1/onnx-mlir/build/src/Runtime/jni/jsoniter/maven/jsoniter-0.9.23.jar D:/a/1/onnx-mlir/build/src/Runtime/jni/jsoniter/jsoniter-0.9.23.jar"
2023-04-15T00:52:21.7814113Z [2/3] Running utility command for javaruntime
2023-04-15T00:52:21.7814696Z [2/3] Running the ONNX-MLIR regression tests
2023-04-15T00:52:21.7815188Z FAILED: test/mlir/CMakeFiles/check-onnx-lit D:/a/1/onnx-mlir/build/test/mlir/CMakeFiles/check-onnx-lit 
2023-04-15T00:52:21.7815986Z cmd.exe /C "cd /D D:\a\1\onnx-mlir\build\test\mlir && C:\hostedtoolcache\windows\Python\3.9.13\x64\python3.exe C:/hostedtoolcache/windows/Python/3.9.13/x64/Scripts/lit.exe -sv D:/a/1/onnx-mlir/build/test/mlir"
2023-04-15T00:52:21.7816661Z ninja: build stopped: subcommand failed.
2023-04-15T00:52:21.8737588Z ##[error]Cmd.exe exited with code '1'.
2023-04-15T00:52:21.9156787Z ##[section]Finishing: Test onnx-mlir
```

**Comment by sorenlassen** [comment]
@gongsu832 thanks for the explanation, but what confuses me is that I cannot pinpoint the problem from reading the contents of the log - can you tell what the cause of the failure is?

**Comment by gongsu832** [comment]
@sorenlassen I pretty much know nothing about lit tests. But from the error message:
```
2023-04-15T00:52:17.4259020Z $ "d:\a\1\onnx-mlir\build\release\bin\onnx-mlir-opt.exe" "-O3" "--convert-krnl-to-affine" "--convert-krnl-to-llvm" "D:\a\1\onnx-mlir\test\mlir\krnl\krnl_getref_lowering.mlir" "-split-input-file"
2023-04-15T00:52:17.4260221Z note: command had no output on stdout or stderr
2023-04-15T00:52:17.4261140Z error: command failed with exit status: 3221226356
```
The `onnx-mlir-opt.exe` failed so it didn't produce anything on stdout, which is supposed to be piped into `filecheck.exe`:
```
2023-04-15T00:52:17.4262373Z $ "d:\a\1\llvm-project\build\bin\filecheck.exe" "D:\a\1\onnx-mlir\test\mlir\krnl\krnl_getref_lowering.mlir"
2023-04-15T00:52:17.4263252Z # command stderr:
2023-04-15T00:52:17.4264126Z FileCheck error: '<stdin>' is empty.
```
So `filecheck.exe` failed. I have no idea why the `onnx-mlir-opt.exe` failed. Perhaps one of the Microsoft folks can help?

**Comment by sorenlassen** [comment]
@jenkins-droid test this please

**Comment by sorenlassen** [comment]
@gongsu832 thank you for looking into this. I never figured it out but I retried the MLIR-Windows-CI once more and it finally succeeded, so it appears to have been a temporary problem of some sort. (It had failed the same way 2 or 3 times when I reported the problem.)

**Pull request merged**